kessel/auth/auth.go

```diff
@@ -0,0 +1,150 @@
+package auth
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/zitadel/oidc/v3/pkg/client"
+)
+
+const expirationWindow = 300  // 5 minutes in second
+const defaultExpiresIn = 3600 // 1 hour in seconds
+
+type OIDCDiscoveryMetadata struct {
+	TokenEndpoint string
+}
+
+type RefreshTokenResponse struct {
+	AccessToken string
+	ExpiresAt   time.Time
+}
+
+type OAuth2ClientCredentials struct {
+	clientId      string
+	clientSecret  string
+	tokenEndpoint string
+	cachedToken   RefreshTokenResponse
+	tokenMutex    sync.Mutex
+}
+
+type FetchOIDCDiscoveryOptions struct {
+	Context    context.Context
+	HttpClient *http.Client
+}
+
+type GetTokenOptions struct {
+	ForceRefresh bool
+	Context      context.Context
+	HttpClient   *http.Client
+}
+
+type oauth2TokenEndpointCaller struct {
+	tokenEndpoint string
+	httpClient    *http.Client
+}
+
+type requestToken struct {
+	ClientID     string `schema:"client_id,omitempty"`
+	ClientSecret string `schema:"client_secret,omitempty"`
+	GrantType    string `schema:"grant_type"`
+}
+
+func MakeOAuth2ClientCredentials(clientId string, clientSecret string, tokenEndpoint string) OAuth2ClientCredentials {
+	return OAuth2ClientCredentials{
+		clientId:      clientId,
+		clientSecret:  clientSecret,
+		tokenEndpoint: tokenEndpoint,
+		cachedToken:   RefreshTokenResponse{},
+		tokenMutex:    sync.Mutex{},
+	}
+}
+
+func FetchOIDCDiscovery(issuerUrl string, options FetchOIDCDiscoveryOptions) (OIDCDiscoveryMetadata, error) {
```

Comment:
The spec does not say anything about FetchOIDCDiscoveryOptions, I think I see why you are doing it though. How does this work with the other sdks, does only Go need this default initialization? Or do we need to update the spec to also allow configuration options?

Reply:
While the spec doesn't say anything about `FetchOIDCDiscoveryOptions` it (the options object) seems to be a common pattern in go to handle optional values. I think i might have missed some commits or updates, because ctx was going to be required, probably forgot to update this part.

Reply:
As discussed, we will keep this as an optional parameter, will update the spec.